chains/evm/contracts/ccvs/VersionedVerifierResolver.sol

```diff
@@ -0,0 +1,109 @@
+  error InvalidDestChainSelector(uint64 destChainSelector);
+  error VersionMismatch(address verifier, bytes4 expected, bytes4 got);
+
+  event InboundImplementationRemoved(bytes4 version);
+  event OutboundImplementationRemoved(uint64 destChainSelector);
+  event InboundImplementationUpdated(bytes4 version, address prevImpl, address newImpl);
+  event OutboundImplementationUpdated(uint64 destChainSelector, address prevImpl, address newImpl);
+
+  struct InboundImplementationArgs {
+    bytes4 version; // ────╮ Verifier version.
+    address verifier; // ──╯ Address of the verifier contract.
+  }
+
+  struct OutboundImplementationArgs {
+    uint64 destChainSelector; // ──╮ Destination chain selector.
+    address verifier; // ──────────╯ Address of the verifier contract.
+  }
+
+  string public constant override typeAndVersion = "VersionedVerifierResolver 1.7.0-dev";
+
+  /// @notice maps verifier versions to their implementation addresses, applied to inbound traffic.
+  mapping(bytes4 version => address verifier) private s_inboundImplementations;
+  /// @notice maps destination chain selectors to their implementation addresses, applied to outbound traffic.
+  mapping(uint64 destChainSelector => address version) private s_outboundImplementations;
+
+  /// @inheritdoc ICrossChainVerifierResolver
+  function getInboundImplementation(
+    bytes calldata ccvData
+  ) external view returns (address) {
+    if (ccvData.length < 4) {
+      return address(0);
+    }
+    return s_inboundImplementations[bytes4(ccvData[:4])];
+  }
+
+  /// @notice Returns the verifier contract for a given version.
+  /// @param version The version of the verifier contract.
+  /// @return verifierAddress The address of the verifier contract.
+  function getInboundImplementationForVersion(
+    bytes4 version
+  ) external view returns (address) {
+    return s_inboundImplementations[version];
+  }
+
+  /// @inheritdoc ICrossChainVerifierResolver
+  function getOutboundImplementation(
+    uint64 destChainSelector
+  ) external view returns (address) {
+    return s_outboundImplementations[destChainSelector];
+  }
+
+  /// @notice Updates inbound implementations.
+  /// @param implementations Verifier versions and their corresponding contracts.
+  function applyInboundImplementationUpdates(
+    InboundImplementationArgs[] calldata implementations
+  ) external onlyOwner {
+    for (uint256 i = 0; i < implementations.length; ++i) {
+      InboundImplementationArgs memory implementation = implementations[i];
+      if (implementation.verifier == address(0)) {
+        // If the verifier address is zero, we clear the implementation for this version.
+        delete s_inboundImplementations[implementation.version];
+        emit InboundImplementationRemoved(implementation.version);
+        continue;
+      }
+      bytes4 expectedVersion = IVersionedVerifier(implementation.verifier).VERSION_TAG();
```

Comment:
Wondering if this is too restrictive. What if we ever want to have a contract that supports multiple versions? Technically this check isn't needed, it's defense in depth so maybe we can ditch it entirely? Also avoids an additional interface